### ISSUE TITLE
Empty window for Simple Counter example

### ISSUE THREAD

**Issue by yay** [maps4print/azul#50]
Hello,

I'm very excited to give this a try and reading the guide ATM:
https://github.com/maps4print/azul/wiki/A-simple-counter

But using the example code above simply shows an empty window without the label and the button:
<img width="912" alt="screen shot 2018-11-10 at 00 52 21" src="https://user-images.githubusercontent.com/1799756/48290135-ec87bc00-e482-11e8-8bcd-4438e3a07305.png">

macOS 10.14.1
Azul version: d39fde59c375c7a3a684a7c90bc8623521351e7e
rustc 1.32.0-nightly (653da4fd0 2018-11-08)

**Comment by fschutt** [comment]
Yes, macOS, especially Mojave is very broken at the moment. This specific bug has to do with https://github.com/tomaka/glutin/issues/1069 and the fix in https://github.com/tomaka/winit/pull/685. More generally, I've tried to debug this in #5, but I don't have a Mac myself, so I can't really debug macOS related issues.

**Comment by fschutt** [comment]
macOS has "depreceated" OpenGL, but hasn't removed it. Being "backend agnostic" is a huge amount of work, you need to redo all the shaders in X different backends. If macOS (idiotically) decides to remove OpenGL (and break all existing apps that rely on it in the process), then webrender is forced to port their shaders to the new backend, if they want to have Firefox be used on that new platform. So no work from my side is necessary, only the shaders in webrender need to be rewritten. Until then, it's going to use OpenGL. There is a Vulkan port of webrender, however, macOS doesn't support Vulkan, so yeah.

No, Azul does not have WebGL, WebGL is internally translated to either DirectX or OpenGL. Allowing users to use their custom rendering API inside of Azul is also a lot of work that is currently not a goal (there are much more pressing issues). However, it would be possible to use DirectComposition on Windows to render DirectX and then OpenGL on top, or to mix Vulkan / OpenGL on Linux.

In terms of performance, these APIs are only useful if you have driver overhead and can parallelize rendering, which is currently the least of Azuls problems. Otherwise, they are not going to make the graphics magically faster. So the only reason to adopt them would be compatibility with platforms, not speed.

I am sorry to say that azul is probably not ready yet on macOS. Besides the winit fixes, the problem seems to be with resizing windows, which has a special API on macOS - Azul does redraw and swap correctly, just macOS doesn't want to display anything.

**Comment by yay** [comment]
Thanks for these insights!

**Comment by fschutt** [comment]
Winit 0.18 with all these fixes was released on crates.io, currently waiting on https://github.com/glium/glium/pull/1711 to update the dependency on winit. So the macOS problems should be fixed soon™.

**Comment by rumanbsl** [comment]
any update on this?

**Comment by fschutt** [comment]
No, the PR has not been merged yet. If you do want to help, contact the glium team. In the future, it is likely that I'll fork glium because the maintenance of it is very unresponsive. However, I haven't had time to do that for now.

**Comment by rumanbsl** [comment]
According to author, glium is not actively developed https://users.rust-lang.org/t/glium-post-mortem/7063, which is a shame :/

**Comment by fschutt** [comment]
Yes, not actively developed **by the author**. There are still other maintainers who merge patches. I've worked with glium for the past year, it's not completely unmaintained.

**Comment by Yatekii** [comment]
Why don't we use gfx-rs?

**Comment by fschutt** [comment]
@rumanbsl @yay Due to inactivity in the glium project, I've now (hopefully only temporarily, but we'll see), forked glium to https://crates.io/crates/azul_glium (and also disabled a few unnecessary dependencies). winit is now the latest version (0.18.0), which should include the patches to close this PR.

I am not sure if this will fix any other Mac-related issues, though, I've only updated the dependencies.

**Comment by yay** [comment]
```
my_first_azul_app $ rustc --version
rustc 1.32.0-nightly (21f268495 2018-12-02)
my_first_azul_app $ cargo build
    Updating git repository `https://github.com/maps4print/azul`
    Updating crates.io index
    Updating git repository `https://github.com/servo/webrender`
error: failed to select a version for `gleam`.
    ... required by package `webrender v0.57.2 (https://github.com/servo/webrender?rev=92235d1fc4ff4c4fb4e1ed709eb3136e96c786b5#92235d1f)`
    ... which is depended on by `azul v0.1.0 (https://github.com/maps4print/azul?rev=ca9f8a2785478ee38c51cf141092b0cdfa0f8057#ca9f8a27)`
    ... which is depended on by `my_first_azul_app v0.1.0 (/Users/vitalykravchenko/projects/my_first_azul_app)`
versions that meet the requirements `^0.6.8` are: 0.6.8

all possible versions conflict with previously selected packages.

  previously selected package `gleam v0.6.5`
    ... which is depended on by `azul v0.1.0 (https://github.com/maps4print/azul?rev=ca9f8a2785478ee38c51cf141092b0cdfa0f8057#ca9f8a27)`
    ... which is depended on by `my_first_azul_app v0.1.0 (/Users/vitalykravchenko/projects/my_first_azul_app)`

failed to select a version for `gleam` which could resolve this conflict
```

**Comment by yay** [comment]
This is when using the latest commit as revision number:
```
[dependencies]
azul = { git = "https://github.com/maps4print/azul", rev = "ca9f8a2785478ee38c51cf141092b0cdfa0f8057" }
```

**Comment by fschutt** [comment]
@yay Whenever this happens, usually it helps to just delete your Cargo.lock file. Sometimes cargo has problems with that when updating an existing lock file.

@asmyers Is there no backtrace?

**Comment by asmyers** [comment]
@fschutt Nope, I set RUST_BACKTRACE=1 and still no backtrace :(

**Comment by yay** [comment]
@fschutt Thanks, that helped. Now I'm at the exact same spot as @asmyers. Backtrace is empty.

**Comment by schulzch** [comment]
I spend some time debugging this, here is a backtrace:
```
azul_glium::context::Context::new::hbfeae991374f7348 (/Users/schulzch/.cargo/registry/src/github.com-1ecc6299db9ec823/azul_glium-0.22.0/src/context/mod.rs:221)
azul_glium::backend::glutin::Display::new_inner::ha4fa6cb9618fe5de (/Users/schulzch/.cargo/registry/src/github.com-1ecc6299db9ec823/azul_glium-0.22.0/src/backend/glutin/mod.rs:113)
azul_glium::backend::glutin::Display::with_debug::h74b1d266f291e1db (/Users/schulzch/.cargo/registry/src/github.com-1ecc6299db9ec823/azul_glium-0.22.0/src/backend/glutin/mod.rs:92)
_$LT$azul..window..Window$LT$T$GT$$GT$::new::h6f9788ed9b59c2c1 (/Users/schulzch/Documents/azul/azul/src/window.rs:776)
_$LT$azul..window..Window$LT$T$GT$$GT$::new_hot_reload::h557030b661e2afab (/Users/schulzch/Documents/azul/azul/src/window.rs:885)
calculator::main::h1b4f08875c4b3919 (/Users/schulzch/Documents/azul/azul/examples/calculator.rs:257)
std::rt::lang_start::_$u7b$$u7b$closure$u7d$$u7d$::hf961f42fdf7a2b95 (@100089ed0..100089ef3:9)
std::rt::lang_start_internal::_$u7b$$u7b$closure$u7d$$u7d$::h38c6d35a8a74d01a (@100d2c090..100d2c0c0:12)
std::panicking::try::do_call::h8a1df18972bca627 (@100d2c090..100d2c0c0:9)
__rust_maybe_catch_panic (@100d37e30..100d37ec0:16)
std::panicking::try::he784830af71690b3 (@100d22570..100d227a0:56)
std::panic::catch_unwind::h17cb4012512d83ac (@100d22570..100d227a0:46)
std::rt::lang_start_internal::h7d0f1ce4d4ec6d26 (@100d22570..100d227a0:46)
std::rt::lang_start::h2dbb14eaca123ceb (@100089e80..100089ed0:20)
main (@1000fd590..1000fd5bc:12)
start (@7fff6877208c..7fff68772095:4)
start (@7fff6877208c..7fff68772095:4)
```
Turns out that this [mod.rs:645](https://github.com/glium/glium/blob/cda2cc0229f430baa72b49a3e407c153221b3c8d/src/context/mod.rs#L645) is triggering this and that it could be related to [glium #1714](https://github.com/glium/glium/issues/1714).

**Comment by asmyers** [comment]
I switched Azul to instantiate the Display using:

    let display = try!(unsafe { Display::unchecked_with_debug(gl_window, DebugCallbackBehavior::Ignore) });

I'm now back to a blank screen with no crashing on my simple hello world application.  Calculator is not crashing and works as well as it did before.

**Comment by schulzch** [comment]
<img width="912" alt="bildschirmfoto 2019-02-09 um 10 19 21" src="https://user-images.githubusercontent.com/2237173/52518977-536c6b00-2c54-11e9-86a1-bfcdc0145923.png">
`cargo run --example calculator` seems to be working in `2d00a019bcce1cf63f11449547db030f723dc44e` on macOS 10.14.2.

(can probably be closed)

**Comment by fschutt** [comment]
Alright, since this seems to be working now on MacOS, I'm closing this - I've also seen apps that run fine on MacOS, so I'm not sure if #5 is still valid. Please reopen if there are still issues.

**Comment by yay** [comment]
![out](https://user-images.githubusercontent.com/1799756/52526814-e26a9900-2ccf-11e9-8944-6aa4ae00e5c8.gif)

- The wrong scale issue at launch only happens when launching on the external (non-retina) monitor (connected to a rMBP).
- The black screen issue happens on both displays.
- Resizing down often results in clipped content.
- Resizing up often results in blank space.

![screen shot 2019-02-10 at 01 07 27](https://user-images.githubusercontent.com/1799756/52526870-6755b280-2cd0-11e9-8dd9-e44bed3dca9e.png)

But hey, it's still progress and much better than a dark window 100% of the time!

**Comment by fschutt** [comment]
@yay Please make a separate issue for your problem, this issue (about the black window at startup) has been solved.

**Comment by schulzch** [comment]
@yay: #5 (HiDPI) or #39 (external displays) would be a good place.
